### DISCUSSION TITLE
Tools to help non compiler engineers optimize their code

### DISCUSSION THREAD

**Discussion by Brian-M-J** [modular/modular#1125] [Ideas]
Mojo is a language that aims to match and maybe exceed the execution speeds of C, C++, Rust, Zig etc ("working backwards from the speed of light "), but it was also mentioned that its compiler would not be magic. I'm assuming this means that the programmer would need to know how to optimize code.

To help them in this, what tools can be provided? The community can build things like profilers and static analysis tools, but things can be made available at the language and compiler level as well. For example, [C++ contracts](https://youtu.be/jTuIzSUxYDc?si=2vZKzrUEH3cDxY8H) have an assume semantic that optimizes code based on some invariants (mentioned at around 35:50). [Wuffs](https://github.com/google/wuffs) uses assertions to guide optimizations. There are also things like [compiler optimization remarks](https://youtu.be/qmEsx4MbKoc?si=hEALFHVndriE-W_S) that hint at possible optimizations.

Edit: Also see #2261. Explainable Profile Guided Optimizations could also be a step to helping programmers optimize their code.

**Comment by david-ragazzi** [comment]
> Mojo is a language that aims to match and maybe exceed the execution speeds of C, C++, Rust, Zig etc ("working backwards from the speed of light "), but it was also mentioned that its compiler would not be magic. I'm assuming this means that the programmer would need to know how to optimize code.
> 
> To help them in this, what tools can be provided? The community can build things like profilers and static analysis tools, but things can be made available at the language and compiler level as well. For example, [C++ contracts](https://youtu.be/jTuIzSUxYDc?si=2vZKzrUEH3cDxY8H) have an assume semantic that optimizes code based on some invariants (mentioned at around 35:50). There are also things like [compiler optimization remarks](https://youtu.be/qmEsx4MbKoc?si=hEALFHVndriE-W_S) that hint at possible optimizations.

Usually IDEs like PyCharm make suggestions over the code or when you ask for a report with improvements.

**Comment by ksandvik** [comment]
There's indeed the possibility of writing tools such as Rust's clippy and integrating this with the VSCode mojo LSP plugin (or a separate one). Now, the language is so young so maybe we need far more iterations (with the protocol/traits, lifetime syntax, object-oriented syntax et rest) before it's worthwhile writing such tools.
